### ISSUE TITLE
Identification of Array Out-of-Bounds Errors in Multiple Locations

### ISSUE THREAD

**Issue by XizheYin-NJU** [rust-lang/regex#1016]
## What version of regex are you using?

```
regex = "=1.8.4"
```

## Describe the bug at a high level.

We have identified four additional errors, which are array out-of-bounds errors in different locations. Please review these errors to determine if they require fixing. The replay files are available at [this link](https://github.com/XizheYin-NJU/regex_replay).

## Error1

Error Message: 
```
thread 'main' panicked at 'range start index 78 out of range for slice of length 16',
/home/yxz/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/regex-1.8.4/src/exec.rs:564:22
```

## Error2
Error Message: 
```
thread 'main' panicked at 'range start index 122 out of range for slice of length 9',
/home/yxz/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/regex-1.8.4/src/exec.rs:512:22
```
## Error3
Error Message: 
```
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 8',
/home/yxz/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/regex-1.8.4/src/re_set.rs:491:1
```
## Error4
```
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 1',
/home/yxz/.cargo/registry/src/mirrors.ustc.edu.cn-61ef6e0cd06fb9b8/regex-1.8.4/src/re_set.rs:462:1
```

**Comment by BurntSushi** [comment]
Why do people keep reporting these? Every single one of these is a result of providing an incorrect offset to functions.

**Comment by XizheYin-NJU** [comment]
I'm sorry for disturbing you. I am using a fuzz testing tool to test the API of this library, which alerts me when the program crashes. I'm not sure if someone has already reported similar errors before. I will pay attention to this issue and provide more useful bug reports for this project in the future.

**Comment by BurntSushi** [comment]
Yes, I understanding you're doing fuzzing. Several people are doing fuzzing and have reported exactly these same issues. One your tests, for example, is giving an incorrect offset to [`SetMatches::matched`](https://docs.rs/regex/latest/regex/struct.SetMatches.html#method.matched). The docs clearly state that the routine will panic if an incorrect offset is given. So crashing in that case is expected behavior. You also have a test for [`Regex::is_match_at`](https://docs.rs/regex/latest/regex/struct.Regex.html#method.is_match_at) which is not documented to panic for an incorrect offset, but that is indeed intended behavior and the docs will be fixed `regex 1.9`.

These fuzz tests are like generating random inputs for `slice` and `i`, and then reporting a bug when `slice[i]` panics.

**Comment by XizheYin-NJU** [comment]
Yes, thank you for your patient response. I will continue to improve my tool in order to discover more valuable issues.
